inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/reorder_weights_opt.cl

```diff
@@ -0,0 +1,160 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "include/common.cl"
+#include "include/data_types.cl"
+
+INIT_INPUT0_INDEX_FUNC_HERE
+INIT_OUTPUT_INDEX_FUNC_HERE
+
+#if OUTPUT_GROUPED
+#   if OUTPUT_DIMS == 5
+#       define IDX_ORDER g, o, i, y, x
+#       define BLOCK_IDX_ORDER g, o_blocked, i_blocked, y, x
+#   elif OUTPUT_DIMS == 6
+#       define IDX_ORDER g, o, i, z, y, x
+#       define BLOCK_IDX_ORDER g, o_blocked, i_blocked, z, y, x
+#   endif
+#else
+#   if OUTPUT_DIMS == 4
+#       define IDX_ORDER o, i, y, x
+#       define BLOCK_IDX_ORDER o_blocked, i_blocked, y, x
+#   elif OUTPUT_DIMS == 5
+#       define IDX_ORDER o, i, z, y, x
+#       define BLOCK_IDX_ORDER o_blocked, i_blocked, z, y, x
+#   endif
+#endif
+#define GET_INDEX(macro, ...) macro(__VA_ARGS__)
+
+#if OUTPUT_TYPE_SIZE == 1
+#   define OUTPUT_BLOCK_WRITE1(ptr, offset, val)   intel_sub_group_block_write_uc1((__global uchar*)(ptr) + (offset), as_uchar(val))
+#   define OUTPUT_BLOCK_WRITE2(ptr, offset, val)   intel_sub_group_block_write_uc2((__global uchar*)(ptr) + (offset), as_uchar2(val))
+#   define OUTPUT_BLOCK_WRITE4(ptr, offset, val)   intel_sub_group_block_write_uc4((__global uchar*)(ptr) + (offset), as_uchar4(val))
+#   define OUTPUT_BLOCK_WRITE8(ptr, offset, val)   intel_sub_group_block_write_uc8((__global uchar*)(ptr) + (offset), as_uchar8(val))
+#elif OUTPUT_TYPE_SIZE == 2
+#   define OUTPUT_BLOCK_WRITE1(ptr, offset, val)   intel_sub_group_block_write_us((__global ushort*)(ptr) + (offset), as_ushort(val))
+#   define OUTPUT_BLOCK_WRITE2(ptr, offset, val)   intel_sub_group_block_write_us2((__global ushort*)(ptr) + (offset), as_ushort2(val))
+#   define OUTPUT_BLOCK_WRITE4(ptr, offset, val)   intel_sub_group_block_write_us4((__global ushort*)(ptr) + (offset), as_ushort4(val))
+#   define OUTPUT_BLOCK_WRITE8(ptr, offset, val)   intel_sub_group_block_write_us8((__global ushort*)(ptr) + (offset), as_ushort8(val))
+#elif OUTPUT_TYPE_SIZE == 4
+#   define OUTPUT_BLOCK_WRITE1(ptr, offset, val)   intel_sub_group_block_write((__global uint*)(ptr) + (offset), as_uint(val))
+#   define OUTPUT_BLOCK_WRITE2(ptr, offset, val)   intel_sub_group_block_write2((__global uint*)(ptr) + (offset), as_uint2(val))
+#   define OUTPUT_BLOCK_WRITE4(ptr, offset, val)   intel_sub_group_block_write4((__global uint*)(ptr) + (offset), as_uint4(val))
+#   define OUTPUT_BLOCK_WRITE8(ptr, offset, val)   intel_sub_group_block_write8((__global uint*)(ptr) + (offset), as_uint8(val))
+#else
+#   error reorder_weights_opt.cl - unsupported output type.
+#endif
```

Comment:
I think you can delete all this macros.

Reply:
Oh, I'm sorry, dont' delete. But maybe to parameterize and simplify as (example for OUTPUT_TYPE_SIZE == 4):

`#   define OUTPUT_BLOCK_WRITE(ptr, offset, val)     CAT(intel_sub_group_block_write, SECOND_BLOCK_SIZE)((__global uint*)(ptr) + (offset), CAT(as_uint, SECOND_BLOCK_SIZE)(val))`

Reply:
Since there already is `data_types.cl` included is there a reason not to use something like:
```c
#define OUTPUT_BLOCK_WRITE(ptr, offset, val) BLOCK_WRITEN(OUTPUT_TYPE, SECOND_BLOCK_SIZE, ptr, offset, val)
```
?
Definitively it will be safer for `OUTPUT_TYPE_SIZE == 1` where `intel_sub_group_block_write_uc` may be unavailable.

Reply:
Thanks, fixed